compiler/AST/AggregateType.cpp

```diff
@@ -2156,6 +2211,131 @@ bool AggregateType::isFieldInThisClass(const char* name) const {
+            !strcmp(psym->name, "BaseRectangularDom")         )
+        && psym->getModule()->modTag == MOD_INTERNAL)
+      return parentAG;
+
+    ag = parentAG; // continue searching
+  }
+
+  return nullptr;  // dummy
+}
+
+
+// Traverse parents until we get to class BaseRectangularDom or
+// BaseArrOverRectangularDom, then return its field 'strides'.
+static Symbol* stridesFieldOfBaseRectParent(AggregateType* ag) {
+  if (AggregateType* parent = baseRectDsiParent(ag))
+    return parent->getField("strides");
+  else
+    return nullptr;
+}
+
+// Returns stridesFieldOfBaseRectParent() if 'use' is in the context
+// of a DSI class
+Symbol* stridesFieldInDsiContext(Expr* use) {
+  if (TypeSymbol* parentSym = toTypeSymbol(use->parentSymbol))
+    if (AggregateType* ag = toAggregateType(parentSym->type))
+      return stridesFieldOfBaseRectParent(ag);
+  return nullptr;
+}
+
+AggregateType* dsiTypeBeingConstructed(CallExpr* parentCall) {
+  if (SymExpr* callee = toSymExpr(parentCall->baseExpr))
+    if (TypeSymbol* calleeTS = toTypeSymbol(callee->symbol()))
+      if (AggregateType* calleeAG =
+          toAggregateType(canonicalClassType(calleeTS->type)))
+        return baseRectDsiParent(calleeAG);
+  return nullptr;
+}
+
+// Same as baseRectDsiParent(), plus returns 'ag'
+// if it is BaseRectangularDom or BaseArrOverRectangularDom,
+static AggregateType* baseRectDsiParentOrSelf(AggregateType* ag) {
+  TypeSymbol* psym = ag->symbol;
+  if (   (!strcmp(psym->name, "BaseArrOverRectangularDom") ||
+          !strcmp(psym->name, "BaseRectangularDom")         )
+      && psym->getModule()->modTag == MOD_INTERNAL)
+    return ag;
+  else
+    return baseRectDsiParent(ag);
+}
+
+// Replaces all uses of stridesFml2 with sblFormal.
+static void replaceStridesWithStridableArg(ArgSymbol* stridesFml2,
+                                           ArgSymbol* sblFormal, bool isBase) {
+  if (isBase) {
+    // For BaseRectangularDom and BaseArrOverRectangularDom, there must be
+    // a single use of stridesFml2 in a call(":", stridesFml2, strideKind).
+    // We replace it with a call("chpl_strideKind", sblFormal).
+    SymExpr* use = stridesFml2->getSingleUse();
+    CallExpr* parent = toCallExpr(use->parentExpr);
+    INT_ASSERT(use, parent->isNamedAstr(astrScolon));
+    parent->replace(new CallExpr("chpl_strideKind", sblFormal));
+
+  } else {
+    for_SymbolSymExprs(se, stridesFml2) {
+      SymExpr* replSE = new SymExpr(sblFormal);
+      // if 'chpl_stridable(strides)', replace all of it with 'stridable'
+      // if 'strides=strides', replace it with 'stridable=stridable'
+      CallExpr* pCall = toCallExpr(se->parentExpr);
+      NamedExpr* pNamed = toNamedExpr(se->parentExpr);
+      if (pCall != nullptr && pCall->isNamed("chpl_stridable"))
+        pCall->replace(replSE);
+      else if (pNamed != nullptr && !strcmp(pNamed->name, "strides"))
+        pNamed->replace(new NamedExpr("stridable", replSE));
+      else
+        se->replace(replSE);
```

Comment:
This isn't a type-safe replacement, is it? If I have a `init` proc that feeds `strides` to something that expects a `strideKind`, and then it gets replaced with a bool, it won't work? Or is there an implicit conversion to bool somehow (seems unlikely to me).

Reply:
This code is invoked only on a clone of the function built by AggregateType::buildDefaultInitializer(). Therefore `strides` is used only when passing to the parent initializer. In this context, the new `stridable` formal is safe to use because the parent initializers will most likely also have overloads that accept `stridable`. Otherwise there will be a compilation failure and the user-defined domain map will need to be ported over. Which is an acceptable outcome.